### PR TITLE
Migrate and clean up more builders / Miscellaneous cleanup

### PR DESCRIPTION
**Pull request by trustin**
Related: #1719

- Add `AppRootFinder` which replaces most factory methods in
  `TomcatService` and `TomcatServiceBuilder`.
- Migrate `CorsServiceBuilder.for*()` to `CorsService.builder()`
- Replace `FileServiceBuilder.for*()` with `FileService.builder()`
- Migrate `HttpFileBuilder.of()` and `ofResource()` to `HttpFile.builder()`
- Migrate `TomcatService.for*()` to `TomcatService.of()`
- Migrate `TomcatServiceBuilder.for*()` to `TomcatService.builder()`
- Migrate `JettyService.for*()` to `JettyService.of()`
- Remove ambiguous factory methods that accepts a single `String`
  parameter.
  - A user must specify a `ClassLoader`, `File` or `Path` now.
  - `FileService.for*()` -> `FileService.of()`
  - `HttpFile.ofResource()` -> `HttpFile.of()`
  - `HttpVfs.ofFileSystem/ClassPath()` -> `HttpVfs.of()`

### CODE REVIEW

**Comment by trustin** [comment]
Added somewhat unrelated changes taking advantage of this mass deprecation PR:

- Deprecations:
  - `DefaultClientBuilderParams` in favor of `ClientBuilderParams.of()`
  - `HttpData.EMPTY_DATA` in favor of `HttpData.empty()`.
  - `DefaultHttpParameters` in favor of `QueryParams`.
  - `DefaultHttpData` in favor of the factory methods in `HttpData`.
    - A user can create an `HttpData` with `endOfStream` by using
      `HttpData.withEndOfStream()` now.
  - `DefaultRpcRequest` in favor of the factory methods in `RpcRequest`.
  - `DefaultRpcResponse` in favor of the factory methods in `RpcResponse`.
  - `NodeValueCodec.DEFAULT` in factor of `NodeValueCodec.ofDefault()`.
- Make `ZooKeeperEndpointGroup` final.
- Remove unused class `MutableValueHolder`

**Comment by trustin** [comment]
@anuraaga Any comments about `HttpData.withEndOfStream()`?

File: core/src/main/java/com/linecorp/armeria/server/ServerPort.java [hunk 1]
```diff
@@ -120,6 +120,7 @@ public InetSocketAddress localAddress() {
      *
      * @deprecated Use {@link #protocols()}.
      */
+    @Nullable
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ServerPort.java]: 👍

**Comment by anuraaga** [comment]
`withEndOfStream` feels a bit weird - unlike headers, I guess it's extremely rare to mutate this flag so it's essentially a factory in the form of `with`. It's also a bit asymmetrical to have it but not e.g., `withData`.

I guess this pattern is to allow setting the flag for the result of the other factories. It's probably ok, but as an alternative, since it's an advanced API what do you think of only providing two factories, `wrapEndOfStream(ByteBuf)` and `wrapEndOfStream(byte[])`?

Anyways I'm ok with either approach

**Comment by trustin** [comment]
I'm not sure I like having additional factory methods, because they would introduce a little bit of inconsistency in the API because some will have `EndOfStream` variants and others won't.

File: core/src/main/java/com/linecorp/armeria/server/cors/CorsService.java [hunk 2]
```diff
@@ -51,6 +56,38 @@
...
+                                  .collect(Collectors.joining(",")));
+            }
+            return builderForAnyOrigin();
+        }
+        return new CorsServiceBuilder(copied);
+    }
+
     private final CorsConfig config;
 
     /**
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/cors/CorsService.java]: We should. :sweat_smile:

File: core/src/main/java/com/linecorp/armeria/common/util/AppRootFinder.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        final Path path = f.toPath();
+        if (isZip(path)) {
+            return path;
+        }
+
+        if (!Files.isDirectory(path)) {
+            throw new IllegalArgumentException(f + " is not a JAR, WAR or directory.");
+        }
+
+        return path;
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/common/util/AppRootFinder.java]: Fixed

**Comment by trustin** [comment]
Thanks for a quick review.

**Pull request merged**